dex/elrond_dex_farm/src/lib.rs

```diff
@@ -219,29 +212,37 @@ pub trait Farm:
         }
 
         let farming_token_id = self.farming_token_id().get();
-        let mut farming_token_amount = amount.clone();
+        let mut initial_farming_token_amount = self.rule_of_three_non_zero_result(
+            &amount,
+            &farm_attributes.current_farm_amount,
+            &farm_attributes.initial_farming_amount,
+        )?;
+        reward += self.rule_of_three(
+            &amount,
+            &farm_attributes.current_farm_amount,
+            &farm_attributes.compounded_reward,
+        );
+
         if self.should_apply_penalty(farm_attributes.entering_epoch) {
             let mut penalty_amount = self.get_penalty_amount(&reward);
             if penalty_amount > 0 {
                 self.burn_tokens(&reward_token_id, &penalty_amount);
                 reward -= penalty_amount;
             }
 
-            penalty_amount = self.get_penalty_amount(&farming_token_amount);
+            penalty_amount = self.get_penalty_amount(&initial_farming_token_amount);
             if penalty_amount > 0 {
                 self.burn_tokens(&farming_token_id, &penalty_amount);
-                farming_token_amount -= penalty_amount;
+                initial_farming_token_amount -= penalty_amount;
```

Comment:
Why isn't there a call to `decrease_farming_token_reserve` here?

Reply:
added

Reply:
Resolved